airbyte-integrations/connectors/destination-snowflake/src/main/java/io/airbyte/integrations/destination/snowflake/typing_deduping/SnowflakeV2TableMigrator.java

```diff
@@ -0,0 +1,115 @@
+    this.databaseName = databaseName;
+    this.generator = generator;
+    this.handler = handler;
+    this.rawNamespace = TypingAndDedupingFlag.getRawNamespaceOverride(RAW_SCHEMA_OVERRIDE).orElse(DEFAULT_AIRBYTE_INTERNAL_NAMESPACE);
+  }
+
+  @Override
+  public void migrateIfNecessary(final StreamConfig streamConfig) throws Exception {
+    final StreamId lowercasedStreamId = buildStreamId_lowercase(
+        streamConfig.id().originalNamespace(),
+        streamConfig.id().originalName(),
+        rawNamespace);
+    final boolean existingTableLowercaseExists = findExistingTable_lowercase(lowercasedStreamId).isPresent();
+    final boolean existingTableUppercaseDoesNotExist = !handler.findExistingTable(streamConfig.id()).isPresent();
+    LOGGER.info(
+        "Checking whether upcasing migration is necessary for {}.{}. Existing lowercased table exists: {}; existing uppercased table does not exist: {}",
+        streamConfig.id().originalNamespace(),
+        streamConfig.id().originalName(),
+        existingTableLowercaseExists,
+        existingTableUppercaseDoesNotExist);
+    if (existingTableLowercaseExists && existingTableUppercaseDoesNotExist) {
+      LOGGER.info(
+          "Executing upcasing migration for {}.{}",
+          streamConfig.id().originalNamespace(),
+          streamConfig.id().originalName());
+      handler.execute(generator.softReset(streamConfig));
+    }
+  }
+
+  // These methods were copied from
+  // https://github.com/airbytehq/airbyte/blob/d5fdb1b982d464f54941bf9a830b9684fb47d249/airbyte-integrations/connectors/destination-snowflake/src/main/java/io/airbyte/integrations/destination/snowflake/typing_deduping/SnowflakeSqlGenerator.java
+  // which is the highest version of destination-snowflake that still uses quoted+lowercased
+  // identifiers
+  private static StreamId buildStreamId_lowercase(final String namespace, final String name, final String rawNamespaceOverride) {
+    // No escaping needed, as far as I can tell. We quote all our identifier names.
+    return new StreamId(
+        escapeIdentifier_lowercase(namespace),
+        escapeIdentifier_lowercase(name),
+        escapeIdentifier_lowercase(rawNamespaceOverride),
+        escapeIdentifier_lowercase(StreamId.concatenateRawTableName(namespace, name)),
+        namespace,
+        name);
+  }
+
+  private static String escapeIdentifier_lowercase(final String identifier) {
+    // Note that we don't need to escape backslashes here!
+    // The only special character in an identifier is the double-quote, which needs to be doubled.
+    return identifier.replace("\"", "\"\"");
+  }
+
+  // And this was taken from
+  // https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connectors/destination-snowflake/src/main/java/io/airbyte/integrations/destination/snowflake/typing_deduping/SnowflakeDestinationHandler.java
+  public Optional<SnowflakeTableDefinition> findExistingTable_lowercase(final StreamId id) throws SQLException {
```

Comment:
nit: is this preserved case rather than lowercase?

Reply:
fixed in https://github.com/airbytehq/airbyte/pull/30068/commits/f5db3aac8b01fb631f06c8db686cab126aad0a29